### PR TITLE
Table: Refactoring state handling for expanded rows

### PR DESCRIPTION
**Pull request by torkelo**
The current code state of Table is too complex, messy, confusing with a mix of useState, useReducer, and 10x useEffect (and a some useCallback)  

It was getting messy even before the expanded rows PR but it made it a bit worse by adding new state handling and passing state functions to getColumns 
https://github.com/grafana/grafana/pull/58682

This PR is just a first step (more refactoring needed!) that tries to simplify the expanded rows feature by using the built in useExpand hook extension for react-table so the expanded indices is stored in the table reducer state. To manage resetting the variable list item height cache I had to add two custom state props to the react table reducer state. 

* Use reducer state instead of useState for expanded Items
* Use table useExpanded hook, which makes the RowExpander a lot simpler 
* Pass table styles everywhere instead of re-building them 
* Some initial refactoring of hooks out from the main Table render function 
* Move reducer to it's own file 

For future PR we should move footer state to reducer state and move those useEffects out of main render.  
After this is merged we should refactor subData prop to use nested data instead of another prop

### CODE REVIEW

File: packages/grafana-ui/src/components/Table/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+  });
+}
+
+/**
+  react-table caches the height of cells so we need to reset them when expanding/collapsing rows
+   We need to take the minimum of the current expanded indexes and the previous expandedIndexes array to account
+  for collapsed rows, since they disappear from expandedIndexes but still keep their expanded height
+ */
+export function useResetVariableListSizeCacheWhenExpendedRowsChange(
```
Comment (codeincarnate) [inline packages/grafana-ui/src/components/Table/hooks.ts:40]: Kind of a nitpick but the name of this hook seems overly long, maybe `useResetVariableListSizeCache` would suffice?

**Review by codeincarnate — APPROVED** [review]
Overall looks good to me and certainly seems like a sensible refactor. I'm a bit on the fence about the hook name 😂, but that still seems pretty minor.

**Review by adrapereira — CHANGES_REQUESTED** [review]
Thanks for the changes! I should have looked further into react-table's internal state when implementing the sub-tables instead of building the state management myself. The changes make sense but I found a few things that should be looked into before merging.

File: packages/grafana-ui/src/components/Table/reducer.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+            onSortByChange(sortByFields);
+          }
+        case 'toggleRowExpanded': {
+          if (action.id) {
+            return {
+              ...newState,
+              lastExpandedIndex: parseInt(action.id, 10),
+              rowExpandedCounter: newState.rowExpandedCounter + 1,
+            };
```
[the remark this answers is not in the document]
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/reducer.ts:55]: @adrapereira I know, the counter is just for the variable list cache effect to trigger

[the remark this answers is not in the document]
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/reducer.ts]: updated the name to match the action it counts

File: packages/grafana-ui/src/components/Table/hooks.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/hooks.ts]: yes!

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -299,45 +197,12 @@ export const Table = memo((props: Props) => {
...
-      listVerticalScrollbarHTML?.remove();
-      (tableScrollbarView as HTMLDivElement).querySelector(':scope > .track-vertical')?.remove();
-
-      (tableScrollbarView as HTMLDivElement).append(listVerticalScrollbarHTML as Node);
-    }
-  });
-
-  useEffect(() => {
-    setExpandedIndexes(new Set());
-  }, [data, subData]);
```
[the remark this answers is not in the document]
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/Table.tsx:336]: Fixed

**Comment by torkelo** [comment]
@adrapereira  thanks for the review, pushed an update that fixed all those comments

File: packages/grafana-ui/src/components/Table/hooks.ts [hunk 4]
```diff
@@ -39,12 +41,13 @@ export function useFixScrollbarContainer(
  */
 export function useResetVariableListSizeCacheWhenExpendedRowsChange(
```
[the remark this answers is not in the document]
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/hooks.ts]: @adrapereira nope, they are both changed

**Review by adrapereira — APPROVED** [review]
Looks great, thanks!

**Pull request merged**